File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
N, M, K = map(int, input().split())

length = 1
while N > length:
    length <<= 1

tree = [0] * (length << 1)
# tree[length] ~ tree[length*2 -1] 까지가 리프 노드

for i in range(N):
    temp = int(input())
    tree[length + i] = temp

# 후위순회하면서 부모노드에 더한값 입력
def postorder(root):
    if root >= length*2:
        return 0
    if tree[root] != 0:
        return tree[root]
    left = postorder(root*2)
    right = postorder(root*2+1)
    tree[root] = left + right
    return tree[root]

postorder(1)

# 구간 합을 구하는 함수
def query_sum(left, right, node, node_left, node_right):
    if right < node_left or node_right < left:
        return 0
    if left <= node_left and node_right <= right:
        return tree[node]
    mid = (node_left + node_right) // 2
    return query_sum(left, right, node*2, node_left, mid) + query_sum(left, right, node*2+1, mid+1, node_right)

for _ in range(M+K):
    a, b, c = map(int, input().split())
    if a == 1:
        temp = (b-1) + length
        diff = c - tree[length+(b-1)]
        while temp != 0:
            tree[temp] += diff
            temp //= 2
    elif a == 2:
        # 구간 b~c의 합을 구하고 출력
        result = query_sum(b-1, c-1, 1, 0, length-1)  # 인덱스를 0-based로 조정
        print(result)

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <iostream>

using namespace std;
typedef long long ll;

ll tree[4000050];
ll arr[1000050];

ll init(int node, int start, int end) {
    if (start == end)
        return tree[node] = arr[start];

    int mid = (start + end) / 2;

    return tree[node] = init(node * 2, start, mid) + init(node * 2 + 1, mid + 1, end);
}

void update(int node, int start, int end, int idx, ll diff) {
    if (idx < start || idx > end) return;

    tree[node] += diff;

    if (start != end) {
        int mid = (start + end) / 2;
        update(node * 2, start, mid, idx, diff);
        update(node * 2 + 1, mid + 1, end, idx, diff);
    }
}

ll sum(int node, int start, int end, int left, int right) {
    if (left > end || right < start) return 0;
    if (left <= start && end <= right) return tree[node];

    int mid = (start + end) / 2;
    return sum(node * 2, start, mid, left, right) + sum(node * 2 + 1, mid + 1, end, left, right);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M, K;
    cin >> N >> M >> K;

    for (int i = 1; i <= N; i++) {
        ll num;
        cin >> num;
        arr[i] = num;
    }

    init(1, 1, N);

    for (int i = 0; i < M + K; i++) {
        ll a, b, c;
        cin >> a >> b >> c;

        // 값 변경
        if (a == 1) {
            ll diff = c - arr[b];
            arr[b] = c;
            update(1, 1, N, b, diff);
        }

        // 구간 합 구하기
        else {
            ll ret = sum(1, 1, N, b, c);
            cout << ret << '\n';
        }
    }
    
    return 0;
}